src/mpi/request/test.c

```diff
@@ -154,11 +134,32 @@ int MPI_Test(MPI_Request *request, int *flag, MPI_Status *status)
     }
 #   endif /* HAVE_ERROR_CHECKING */
 
+    /* If this is a null request handle, then return an empty status */
+    if (*request == MPI_REQUEST_NULL) {
+        MPIR_Status_set_empty(status);
+        *flag = TRUE;
+        goto fn_exit;
+    }
+
     /* ... body of routine ...  */
 
     mpi_errno = MPIR_Test_impl(request, flag, status);
     if (mpi_errno) goto fn_fail;
     
+    if (MPIR_Request_is_complete(request_ptr)) {
+        mpi_errno = MPIR_Request_complete(request, request_ptr, status,
+                                          &active_flag);
```

Comment:
`MPIR_Request_complete` frees the MPIR request object, sets `MPI_REQUEST_NULL` to the user request, and sets the status and the flag.

This may be something we want to consider renaming. The current naming of these functions are confusing. How about `MPIR_Request_post_complete` for something?

Reply:
This is the same as the previous code, we have to check if it is really completed first then call `Request_complete` which really does the cleanup.

Reply:
@hajimefu `post_complete` is an awful name.  It doesn't give any useful information.

Besides, I don't think we need to complete the request again.  In fact, I'd argue that you should *not* complete the request again (since the `cc` has already hit zero).  You should simply free it.

Reply:
I'd recommend a separate patch for cleaning that up as a prereq patch, instead of merging that piece with this patch.

Reply:
@pavanbalaji I have to admit I'm not good at naming...
Before going into that discussion, I want to make it clear that `MPIR_Request_complete` does *NOT* complete the request (=it does not decrement the completion counter). I think this is where all the confusion comes from, and this is why I'm proposing renaming it.

Reply:
Oof.  `MPID_Request_complete` decrements the `cc`.  I agree that the name is just awful and should be renamed.

Reply:
Okay. Now we are in the same page. Do you want to come up with a better name? `MPIR_Request_cleanup`?

Reply:
In fact, the only thing it's not doing is "completing" the request.  :-)

Reply:
`MPIR_Request_finalize` ? If nothing better I will just leave it to move on first.